src/Ouroboros/Network/MonadClass/MonadTimer.hs

```diff
@@ -27,12 +39,72 @@ mult n = sum . replicate n
 fromStart :: TimeMeasure t => Duration t -> t
 fromStart = flip addTime zero
 
-class (Monad m, TimeMeasure (Time m)) => MonadTimer m where
-  type Time m :: *
-  timer :: Duration (Time m) -> m () -> m ()
+data TimeoutState = TimeoutPending | TimeoutFired | TimeoutCancelled
+
+class (MonadSTM m stm, TimeMeasure (Time m)) => MonadTimer m stm where
+  type Time    m :: *
+  data Timeout m :: *
+
+  -- | Create a new timeout which will fire at the given time duration in
+  -- the future.
+  --
+  -- The timeout will start in the 'TimeoutPending' state and either
+  -- fire at or after the given time leaving it in the 'TimeoutFired' state,
+  -- or it may be cancelled with 'cancelTimeout', leaving it in the
+  -- 'TimeoutCancelled' state.
+  --
+  -- Timeouts /cannot/ be reset to the pending state once fired or cancelled
+  -- (as this would be very racy). You should create a new timeout if you need
+  -- this functionality.
+  --
+  newTimeout     :: Duration (Time m) -> m (Timeout m)
+
+  -- | Read the current state of a timeout. This does not block, but returns
+  -- the current state. It is your responsibility to use 'retry' to wait.
+  --
+  -- Alternatively you may wish to use the convenience utility 'awaitTimeout'
+  -- to wait for just the fired or cancelled outcomes.
+  --
+  -- You should consider the cancelled state if you plan to use 'cancelTimeout'.
+  --
+  readTimeout    :: Timeout m -> stm TimeoutState
+
+  -- Adjust when this timer will fire, to the given duration into the future.
+  --
+  -- It is safe to race this concurrently against the timer firing. It will
+  -- have no effect if the timer fires first.
+  --
+  -- The new time can be before or after the original expiry time, though
+  -- arguably it is an application design flaw to move timeouts sooner.
+  --
+  updateTimeout  :: Timeout m -> Duration (Time m) -> m ()
+
+  -- | Cancel a timeout (unless it has already fired), putting it into the
+  -- 'TimeoutCancelled' state. Code reading and acting on the timeout state
+  -- need to handle such cancellation appropriately.
+  --
+  -- It is safe to race this concurrently against the timer firing. It will
+  -- have no effect if the timer fires first.
+  --
+  cancelTimeout  :: Timeout m -> m ()
+
+  -- | Returns @True@ when the timeout is fired, or @False@ if it is cancelled.
+  awaitTimeout   :: Timeout m -> stm Bool
+  awaitTimeout t  = do s <- readTimeout t
+                       case s of
+                         TimeoutPending   -> retry
+                         TimeoutFired     -> return True
+                         TimeoutCancelled -> return False
+
+  threadDelay    :: Duration (Time m) -> m ()
+  threadDelay d   = void . atomically . awaitTimeout =<< newTimeout d
+
+{-# DEPRECATED timer "Use threadDelay or the new timer API instead" #-}
+timer :: MonadTimer m stm => Duration (Time m) -> m () -> m ()
+timer d action = fork (threadDelay d >> action)
```

Comment:
@dcoutts it might be worth adding `registerDelay` which I belive we can implement using this api and provide `IO` implementation by [registerDelay](https://hackage.haskell.org/package/stm-2.5.0.0/docs/Control-Concurrent-STM-TVar.html#v:registerDelay)  Andrzej used `registerDelay` to implement simple timers that run in `cardano-sl`.